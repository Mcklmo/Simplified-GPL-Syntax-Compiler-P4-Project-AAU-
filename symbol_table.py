class SymbolTable:
    def __init__(self, vars,functions,childs):
        self.vars = vars
        self.functions = functions
        self.childs = childs

class Var:
    def __init__(self, name, type, value=None):
        self.name = name
        self.type = type
        # value is used to type check the variable in the type checking compiler phase
        # it is defaulted to none because Vars are used in function parameters
        self.value = value

class Function:
    def __init__(self, name, type, params, block):
        self.name = name
        self.type = type
        self.params = params
        self.block = block