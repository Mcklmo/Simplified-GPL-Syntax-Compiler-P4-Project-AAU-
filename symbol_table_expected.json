{
  "vars": [
    {
      "a": {
        "type": "num",
        "value": 0
      }
    }
  ],
  "functions": [
    {
      "b": {
        "type": "string",
        "params": [
          {
            "s": {
              "type": "string"
            }
          }
        ],
        "block": {
          "vars": [
            {
              "a": {
                "type": "num",
                "value": 1
              }
            }
          ],
          "functions": [],
          "childs": [
            {
              "vars": [],
              "functions": [],
              "childs": [
                {
                  "vars": [
                    {
                      "b": {
                        "type": "num",
                        "value": 9
                      }
                    }
                  ],
                  "functions": [],
                  "childs": []
                }
              ]
            }
          ]
        }
      }
    }
  ],
  "childs": [
    {
      "vars": [
        {
          "c": {
            "type": "num",
            "value": 8
          }
        }
      ],
      "functions": [],
      "childs": []
    }
  ]
}
